File: lev1/문자열 다루기 기본.py
"""
문제 설명
문자열 s의 길이가 4 혹은 6이고, 숫자로만 구성돼있는지 확인해주는 함수, solution을 완성하세요.
예를 들어 s가 a234이면 False를 리턴하고 1234라면 True를 리턴하면 됩니다.

제한 사항
s는 길이 1 이상, 길이 8 이하인 문자열입니다.

입출력 예
s	return
a234	false
1234	true
"""


def solution(s):
    # 0.003154566666666671
    # return s.isdigit() if len(s) == 4 or len(s) == 6 else False

    # code refactoring 요소
    # 0.0023687666666666676
    return s.isdigit() and (len(s) == 4 or len(s) == 6)

print(solution("a234"))
print(solution("1234"))
print(solution("1"))

import timeit
avg_time = 0.
li = ['a234', '1234', '1']
for n in li:
    avg_time += timeit.timeit(lambda: solution(n), number=10000)
print(f'avg_time: {avg_time / len(li)}')

File: lev1/문자열_다루기_기본.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <ctype.h>

// 파라미터로 주어지는 문자열은 const로 주어집니다. 변경하려면 문자열을 복사해서 사용하세요.
bool solution(const char* s) {
    // // 0.000830
    // int N = strlen(s);
    // int zero = (int)'0';

    // if(N == 4 || N == 6){
    //     for(int i=0; i<N; i++){
    //         if(zero>s[i] || zero+9<s[i])
    //             return false;
    //     }
    //     return true;
    // }
    // return false;

    // using ctype-digit function - 0.001284
    int N = strlen(s);
    if(N == 4 || N == 6){
        for(int i=0; i<N; i++){
            if(!isdigit(s[i]))
                return false;
        }
        return true;
    }
    return false;
}


int main(){
    clock_t tic = clock();
    for (int i=0; i<10000; i++){
        solution("a234");
        solution("1234");
        solution("01234567");
        solution("012347");
        solution("0ZZZ");
        solution("0123Z7");
        solution("654321");
    }
    clock_t toc = clock();
    printf("running time: %f\n", (double)(toc-tic) / CLOCKS_PER_SEC);
    return 0;
}
